src/net_processing.cpp

```diff
@@ -4907,6 +4907,7 @@ bool PeerManagerImpl::ProcessMessages(CNode* pfrom, std::atomic<bool>& interrupt
             LOCK(peer->m_getdata_requests_mutex);
             if (!peer->m_getdata_requests.empty()) fMoreWork = true;
         }
+        if (m_orphanage.HaveTxToReconsider(peer->m_id)) fMoreWork = true;
```

Comment:
This doesn't work after the other changes in this PR -- you might be creating more work for some other peer, so should be checking if any peers have txs to reconsider, and only doing this once after looping through all the peers.

Reply:
So, I guess this raises the question: should we move orphan processing outside of `ProcessMessages` and `SendMessages` entirely, and just do it separately, independent of both who sent the orphan and who sent the parent?

Unless someone has a good reason not to, I think I might draft that up as an alternative PR so the approaches can be compared more directly.

Reply:
> should we move orphan processing outside of ProcessMessages and SendMessages entirely, and just do it separately, independent of both who sent the orphan and who sent the parent?

Had a similar thought, mainly I don't see why we should try to process all orphans before moving on to `ProcessMessage()`... afaict we never don't a message to a peer as a result of processing an orphan (other than potentially disconnecting) so maintaining the order of responses is not justification like it is for finishing `ProcessGetData()` queue.

fwiw doing orphan processing independently seems reasonable to me, though I don't have a very strong grasp of all the possible implications.

Reply:
I thought of it more along the lines of "if a node gives us a lot of orphans to process, and we're now processing them, we shouldn't do more work for that node until all the orphan work is done; also we shouldn't delay work from other nodes while we're processing that node's orphans".

Reply:
> So, I guess this raises the question: should we move orphan processing outside of `ProcessMessages` and `SendMessages` entirely,

Where do you intend to move it? Into a dedicated thread for orphan processing?

> This doesn't work after the other changes in this PR -- you might be creating more work for some other peer, so should be checking if any peers have txs to reconsider, and only doing this once after looping through all the peers.

Is this situation (that the peer who sent us an orphan is different from the peer that sent us the parent) typical? Otherwise might it be acceptable to just have a unnecessary 100ms wait if it occurs very rarely?

Reply:
> Where do you intend to move it? Into a dedicated thread for orphan processing?

No, I was thinking a separate `ProcessOrphans` method like `ProcessMessages` and `SendMessages`.

> Is this situation (that the peer who sent us an orphan is different from the peer that sent us the parent) typical? Otherwise might it be acceptable to just have a unnecessary 100ms wait if it occurs very rarely?

On a long running node, I'm seeing maybe a few hundred orphans being accepted during startup (ie, children whose parent txs were announced while my node was offline) but then perhaps only 20-60 accepted orphans per day. I don't have much of an idea how many of those would have a different node sending the parent compared to the child, but it doesn't seem likely to be a real problem.

Reply:
> should we move orphan processing outside of ProcessMessages and SendMessages entirely, and just do it separately, independent of both who sent the orphan and who sent the parent?
> 
> Unless someone has a good reason not to, I think I might draft that up as an alternative PR so the approaches can be compared more directly.

See #19364